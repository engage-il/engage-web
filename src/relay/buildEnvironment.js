// @flow
import { DefaultNetworkLayer, Environment } from 'react-relay/classic'
import config from '../config'

export function buildEnvironment (headers: Object = {}): any {
  // recreate relay environment
  const environment = new Environment()
  const networkLayer = new DefaultNetworkLayer(config.graphUrl, { headers })
  environment.injectNetworkLayer(networkLayer)
  return environment
}
