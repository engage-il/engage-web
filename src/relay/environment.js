// @flow
import Relay from 'react-relay'
import config from '../config'

export function createRelayEnvironment () {
  // NOTE: this becomes a bit more complex when we support sign-in/out
  Relay.injectNetworkLayer(
    new Relay.DefaultNetworkLayer(config.graphUrl)
  )
}
