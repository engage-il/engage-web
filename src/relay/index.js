// @flow
export { currentEnvironment } from './currentEnvironment'
