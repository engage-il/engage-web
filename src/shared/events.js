// @flow
import EventEmitter from 'eventemitter3'

class Events extends EventEmitter {
  setAuthHeader = '@@events/set-auth-header'
  setEnvironment = '@@events/set-environment'
  showNotification = '@@events/show-notification'
}

export const events = new Events()
