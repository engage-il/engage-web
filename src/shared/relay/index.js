// @flow
export { currentEnvironment } from './currentEnvironment'
export { unwrap } from './unwrap'
export { withLoadMoreArgs } from './withLoadMoreArgs'
export { cacheResolvers } from './cacheResolvers'
export { createPaginationCacheResolver } from './createPaginationCacheResolver'
