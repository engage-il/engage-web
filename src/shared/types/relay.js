// @flow
import type { ConcreteBatch, QueryPayload, QueryResponseCache } from 'relay-runtime'

export type RelayCacheResovler = {
  canCacheRequest (
    operation: ConcreteBatch,
    variables: Object
  ): boolean,
  getCachedResponse (
    operation: ConcreteBatch,
    variables: Object,
    cache: QueryResponseCache
  ): ?QueryPayload,
  setCachedResponse (
    operation: ConcreteBatch,
    variables: Object,
    response: QueryPayload,
    cache: QueryResponseCache
  ): void
}
