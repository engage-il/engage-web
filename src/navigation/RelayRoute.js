// @flow
import React, { Component } from 'react'
import { QueryRenderer } from 'react-relay'
import { Route } from 'react-router-dom'
import type { ContextRouter } from 'react-router-dom'
import { currentEnvironment, cacheResolvers } from 'shared/relay'
import { events } from 'shared/events'
import type { RelayRouteConfig } from 'shared/types'

export class RelayRoute extends Component {
  props: {
    path: string,
    exact?: boolean
  } & RelayRouteConfig

  render () {
    const { props } = this
    const { path, exact, ...rest } = props

    // ...rest not playing nicely with flow for some reason
    const config = {
      ...rest,
      query: props.query,
      render: props.render
    }

    return <Route
      path={path}
      exact={exact}
      component={withRelay(config)}
    />
  }
}

const withRelay = (config: RelayRouteConfig) => (
  class Container extends Component {
    props: ContextRouter

    state: { environment: Object } = {
      environment: currentEnvironment()
    }

    // events
    didSetEnvironment = () => {
      this.setState({ environment: currentEnvironment() })
    }

    // lifecycle
    componentWillMount () {
      events.on(events.setEnvironment, this.didSetEnvironment)

      const { cacheResolver } = config
      if (cacheResolver) {
        cacheResolvers.add(cacheResolver)
      }
    }

    componentWillUnmount () {
      events.off(events.setEnvironment, this.didSetEnvironment)

      const { cacheResolver } = config
      if (cacheResolver) {
        cacheResolvers.remove(cacheResolver)
      }
    }

    render () {
      const { environment } = this.state
      const { query, getInitialVariables, render } = config
      const { params } = this.props.match

      // merge config variables and route variables
      const initialVariables = getInitialVariables && getInitialVariables(this.props)
      const variables = {
        ...initialVariables,
        ...params
      }

      return <QueryRenderer
        environment={environment}
        query={query}
        variables={variables}
        render={({ error, props }: { error: ?Error, props: ?Object }) => {
          if (error) {
            throw new Error(error) // TODO: show error view
          }

          return render(props)
        }}
      />
    }
  }
)
