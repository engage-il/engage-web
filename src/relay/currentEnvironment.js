// @flow
import { Environment, RecordSource, Store, Network } from 'relay-runtime'
import config from '../config'
import { events } from 'shared/events'

// current environment
let environment = createEnvironment()

function didSetAuthHeader (header: string) {
  environment = createEnvironment({ 'Authorization': header })
  events.emit(events.setEnvironment)
}

export function currentEnvironment (): Object {
  return environment
}

events.on(events.setAuthHeader, didSetAuthHeader)

// helpers
// see: https://facebook.github.io/relay/docs/network-layer.html
function createQuery (headers: Object) {
  return async function query (operation, variables): Object {
    const response = await fetch(config.graphUrl, {
      method: 'POST',
      headers: {
        'content-type': 'application/json',
        ...headers
      },
      body: JSON.stringify({
        query: operation.text,
        variables
      })
    })

    return response.json()
  }
}

// see: https://facebook.github.io/relay/docs/relay-environment.html
function createEnvironment (headers: Object = {}): Object {
  const source = new RecordSource()
  const store = new Store(source)
  const network = Network.create(createQuery(headers))

  return new Environment({ store, network })
}
