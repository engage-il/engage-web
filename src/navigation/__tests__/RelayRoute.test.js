/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'
import { RelayRoute } from '../RelayRoute'
import { events } from 'shared/events'
import { currentEnvironment, cacheResolvers } from 'shared/relay'
import { routerProps } from 'mocks/routerProps'

// mocks
jest.mock('shared/relay', () => ({
  currentEnvironment: jest.fn(),
  cacheResolvers: {
    add: jest.fn(),
    remove: jest.fn()
  }
}))

// subject
let subject
let container
let path
let exact
let config = {}

function loadSubject () {
  subject = shallow(<RelayRoute path={path} exact={exact} {...config} />)
}

function loadContainer () {
  loadSubject()
  const Container = element.route().prop('component')
  container = shallow(<Container {...routerProps} />)
}

const element = {
  route: () => subject.find('Route'),
  renderer: () => container.find('QueryRenderer')
}

// spec
afterEach(() => {
  container && container.unmount()
  subject = null
  container = null
})

describe('#render', () => {
  it('propogates the path', () => {
    path = '/test/path'
    loadSubject()
    expect(element.route()).toHaveProp('path', path)
  })

  it(`propogates 'exact'`, () => {
    exact = true
    loadSubject()
    expect(element.route()).toHaveProp('exact', exact)
  })
})

describe('the container', () => {
  describe('#state', () => {
    it('defaults to the current environment', () => {
      const environment = { foo: 'bar' }
      currentEnvironment.mockReturnValueOnce(environment)
      loadContainer()
      expect(container).toHaveState('environment', environment)
    })
  })

  describe('#componentWillMount', () => {
    it('adds the cache resolver if it exists', () => {
      config = { cacheResolver: { foo: 'bar' } }
      loadContainer()
      expect(cacheResolvers.add).toHaveBeenCalledWith(config.cacheResolver)
    })
  })

  describe('#render', () => {
    it('propogates the query', () => {
      config = { query: 'foo' }
      loadContainer()
      expect(element.renderer()).toHaveProp('query', config.query)
    })

    it('combines the initial config variables and route params', () => {
      const params = { bar: 'baz' }
      const variables = { foo: 'bar' }

      routerProps.match.params = params
      config = { getInitialVariables: () => variables }
      loadContainer()

      expect(element.renderer()).toHaveProp('variables', {
        ...variables,
        ...params
      })
    })

    it('delegates to the config render when there is no error', () => {
      const props = { foo: 'bar' }
      config = { render: jest.fn() }
      loadContainer()

      const render = element.renderer().prop('render')
      render({ error: null, props })

      expect(config.render).toHaveBeenCalledWith(props)
    })
  })

  describe('when environment changes', () => {
    it('re-renders with the new environment', () => {
      loadContainer()
      const environment = { bar: 'baz' }
      currentEnvironment.mockReturnValueOnce(environment)

      events.emit(events.setEnvironment)
      expect(element.renderer()).toHaveProp('environment', environment)
    })
  })
})
