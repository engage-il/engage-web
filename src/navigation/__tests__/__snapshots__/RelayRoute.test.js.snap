// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#render propogates 'exact' 1`] = `
<Route
  component={[Function]}
  exact={true}
  path="/test/path"
/>
`;

exports[`#render propogates the path 1`] = `
<Route
  component={[Function]}
  path="/test/path"
/>
`;
